pkg/threescale/backend/manager.go

```diff
@@ -0,0 +1,367 @@
+	appID := transaction.Params.UserKey
+
+	if appID != "" {
+		return appID
+	}
+	return transaction.Params.AppID
+}
+
+// handleCacheMiss will attempt to call remote 3scale with a blank request and the required
+// extensions enabled to learn current state for caching purposes.
+// Sets the value in the cache and returns the newly cached value for re-use if required
+func (b *Backend) handleCacheMiss(request threescale.Request, cacheKey string) (Application, error) {
+	var app Application
+	emptyTransaction, err := b.blankRequestFromTransactions(request.Transactions)
+	if err != nil {
+		return app, err
+	}
+
+	emptyRequest := threescale.Request{
+		Auth:       request.Auth,
+		Extensions: api.Extensions{api.HierarchyExtension: "1"},
+		Service:    request.Service,
+		Transactions: []api.Transaction{
+			emptyTransaction,
+		},
+	}
+	resp, err := b.remoteAuth(emptyRequest)
+	if err != nil {
+		return app, err
+	}
+	app = b.applicationFromResponse(resp)
+
+	b.cache.Set(cacheKey, app)
+	return app, nil
+}
+
+func (b *Backend) blankRequestFromTransactions(transactions []api.Transaction) (api.Transaction, error) {
+	if transactions == nil || len(transactions) == 0 {
+		return api.Transaction{}, fmt.Errorf("transaction must be non-nil and non-empty")
+	}
+
+	transactionCopy := make([]api.Transaction, len(transactions))
+	copy(transactionCopy, transactions)
+	transactionCopy[0].Metrics = nil
+
+	return transactionCopy[0], nil
+}
+
+func (b *Backend) applicationFromResponse(resp threescale.AuthorizeResult) Application {
+	limitCounter := make(map[string]map[api.Period]*Limit)
+
+	for metric, report := range resp.GetUsageReports() {
+		l := &Limit{
+			current: report.CurrentValue,
+			max:     report.MaxValue,
+		}
+		limitCounter[metric] = make(map[api.Period]*Limit)
+		limitCounter[metric][report.PeriodWindow.Period] = l
+	}
+
+	return Application{
+		LastResponse:  resp,
+		LimitCounter:  limitCounter,
+		UnlimitedHits: make(map[string]int),
+	}
+}
+
+func (b *Backend) computeAffectedMetrics(hierarchy api.Hierarchy, metrics api.Metrics) api.Metrics {
+	for parent, children := range hierarchy {
+		for k, v := range metrics {
+			if contains(k, children) {
+				if _, known := metrics[parent]; known {
+					metrics[parent] += v
+				} else {
+					metrics.Add(parent, v)
+				}
+			}
+		}
+	}
+	return metrics
+}
+
+// isAuthorized works on a copy of the passed Application and will inform the caller if limits have
+// been breached in the provided affected metrics
+func (b *Backend) isAuthorized(application Application, affectedMetrics api.Metrics) (Application, bool) {
+	copiedApplication := application.deepCopy()
+	commitChanges := true
+
+out:
+	for metric, incrementBy := range affectedMetrics {
+		cachedValue, ok := copiedApplication.LimitCounter[metric]
+		if ok {
+			for _, limit := range cachedValue {
+				newValue := limit.current + incrementBy
+				if newValue > limit.max {
+					commitChanges = false
+					break out
```

Comment:
Doesn't this restart the loop again? The label `out` would be more suited to exit completely the loop, ie. be placed right before the return statement (in which case a label is probably not required).

Also, this would be more like internal semantics of how we want this function to behave, but if we exit early here then the returned application cannot be used for anything in case the boolean result is false (which is ok to do and preferred for those cases where we don't care about the status of all these metrics if it's going to be unauthorised anyway).

Reply:
No this exits the loop entirely.

Reply:
Ok, `out` states the loop from which you want to exit. :+1: